src/core/annotation.js

```diff
@@ -1367,6 +1375,71 @@ class WidgetAnnotation extends Annotation {
     }
     return localResources || Dict.empty;
   }
+
+  _collectJS(entry, xref, list, parents) {
+    if (!entry) {
+      return;
+    }
+
+    let parent = null;
+    if (isRef(entry)) {
+      if (parents.has(entry)) {
+        // If we've already met entry then we've a cycle
+        return;
+      }
+      parent = entry;
+      parents.add(parent);
+      entry = xref.fetch(entry, true);
+    }
+    if (Array.isArray(entry)) {
+      for (const e of entry) {
+        this._collectJS(e, xref, list, parents);
+      }
+    } else if (entry instanceof Dict) {
+      if (entry.get("S") === Name.get("JavaScript") && entry.has("JS")) {
+        const code = stringToPDFString(entry.get("JS"));
+        if (code) {
+          list.push(code);
+        }
+      }
+      this._collectJS(entry.getRaw("Next"), xref, list, parents);
+    }
+
+    if (parent) {
+      parents.delete(parent);
+    }
+  }
+
+  _collectActions(xref, dict) {
+    const actions = Object.create({});
+    if (dict.has("AA")) {
+      const additionalAction = dict.get("AA");
+      for (const key of additionalAction.getKeys()) {
+        if (key in EVENT_MAP) {
+          const actionDict = additionalAction.getRaw(key);
+          const parents = new Set();
```

Comment:
Given how this is actually used in `_collectJS`, for storing of `Ref`s, you really want a `RefSet` instead here to prevent issues related to caching.